tests/unit/test_util.py

```diff
@@ -106,17 +102,6 @@ def test_import_optional_dependency(self):
             pybtex = sys.modules["pybtex"]
             sys.modules["pybtex"] = None
             pybamm.print_citations()
-        with self.assertRaisesRegex(
-            ModuleNotFoundError, "Optional dependency anytree is not available."
-        ):
-            with TemporaryDirectory() as dir_name:
-                sys.modules["anytree"] = None
-                test_stub = os.path.join(dir_name, "test_visualize")
-                test_name = f"{test_stub}.png"
-                c = pybamm.Variable("c", "negative electrode")
```

Comment:
We can also move the file outside `tests/` and into the `scripts/` folder, and directly run it if we don't want it to be collected by `unittest` – it is just a single file, after all.

Reply:
I agree, moving the module to `/scripts` and renaming it `import_pybamm_test.py` sounds good, but shouldn't the `test_import_optional_dependency` function be retained to ensure that `import_optional_dependency` keep working correctly despite (eventual) changes? So if a dependency is not available it must throw an exception, otherwise import it correctly.

Reply:
The `test_import_optional_dependency` function was testing whether PyBaMM could be imported without any optional dependencies (by mocking the optional dependencies as if they were not installed), but I assume the same behaviour is being tested by `tests/dependencies.py` right now?

P.S. Another way that I thought about doing this is to move the contents of `tests/dependencies.py` to `test_util.py` like before, and somehow make `unitttest` not collect these tests by default during `nox -s unit` / `python run-tests.py --unit` and therefore not run them unless specifically pointed to do so (through the `python -m unittest tests.test_util.<...>` command or similar). Maybe there is a decorator or some functionality in `pytest` that can do this – i.e., ignore a test X when running all tests (X, Y, Z) but do run X if it is the only test collected.

Reply:
> The `test_import_optional_dependency` function was testing whether PyBaMM could be imported without any optional dependencies (by mocking the optional dependencies as if they were not installed), but I assume the same behaviour is being tested by `tests/dependencies.py` right now?

The current idea I have is that `test_import_optional_dependency`, being a unit test, should only test that the `import_optional_dependency` does its job as expected, i.e. that is able to import a library if it is available (otherwise it must fail). 
The `tests/dependencies.py` module, on the other hand, should test that:
 - `test_core_optional_dependencies` - `pybamm` core does not contain more libraries than we expect (e.g. optional ones).
 - `test_core_pybamm_import` - `pybamm` core is able to run "on its own" (it does not rely on optional dependencies, like instead `pybamm[all]`)
 
 But this is just how I thought it, let me know if it does make sense for you.

> P.S. Another way that I thought about doing this is to move the contents of `tests/dependencies.py` to `test_util.py` like before, and somehow make `unitttest` not collect these tests by default during `nox -s unit` / `python run-tests.py --unit` and therefore not run them unless specifically pointed to do so (which would be through the `python -m unittest tests.test_util.<...>` command or similar). Maybe there is a decorator in `unittest` or some functionality in `pytest` that can do this – i.e., ignore a test X when running all tests (X, Y, Z) but do run X if it is the only test collected.

Great idea, it was my first strategy too, but I couldn't figure out how to ignore single tests with `unittest`/`pytest` when all tests are launched with `nox`. Maybe we could edit `noxfile.py` and/or `run_tests.py` to make it work (?)

Reply:
> The current idea I have is that `test_import_optional_dependency`, being a unit test, should only test that the `import_optional_dependency` does its job as expected, i.e. that is able to import a library if it is available (otherwise it must fail). The `tests/dependencies.py` module, on the other hand, should test that:
> 
> * `test_core_optional_dependencies` - `pybamm` core does not contain more dependencies than we expect (e.g. optional ones hidden inside it, like `pandas`).
> * `test_core_pybamm_import` - `pybamm` core is able to run "on its own" (it does not rely on optional dependencies, like instead `pybamm[all]`)

IMO, the `test_import_optional_dependency` function should test whether `import_optional_dependency` is able to raise the appropriate `ModuleNotFoundError` if an optional dependency is not found to be installed, and not whether it can import an optional dependency if it is installed (whether it is able to _import_ said dependency is already being tested in the tests for `importlib.import_module()` in the upstream CPython/stdlib repository, because that's what we are using under the hood). To find these optional dependencies for PyBaMM, we can use the `importlib.metadata.metadata.json['<...>']` attribute. So I guess the issue is that this attribute is available on Python 3.10 and later, which restricts the use of it on Python 3.8 and Python 3.9, which is also why we are testing this on a separate workflow with Python 3.11 installed and in a separate file that doesn't get called during the regular unit tests.

A solution is as follows: if we can edit `noxfile.py` and install the backport package `importlib-metadata` in the `unit` session (see [the documentation for `importlib_metadata`](https://importlib-metadata.readthedocs.io/en/latest/api.html) – there might be a few differences in the API), it will give us this API for Python 3.8 and Python 3.9 too, with which we can bring both `test_core_pybamm_import` and `test_core_optional_dependencies` back to `test_util.py` instead of the separate file that has been created.

> Great idea, it was my first strategy too, but I couldn't figure out how to ignore single tests with `unittest`/`pytest` when all tests are launched with `nox`. Maybe we could edit `noxfile.py` and/or `run_tests.py` to make it work (?)

Yes. With `importlib_metadata`, the scenario could possibly look like this:

1. Edit `noxfile.py` to install it for `nox -s unit`, while installing PyBaMM with all optional dependencies as it currently does
2. Use `importlib_metadata` to access and parse available dependencies from PyBaMM's dependency tree (except the `[dev]` and `[docs]` extras, of course)
3. Mock all of these dependencies (set them to `None` in `sys.modules`), and then
    a. Test whether `import_optional_dependency` raises the warning/error it is supposed to raise for all of these dependencies
    b. Test whether core PyBaMM can be imported without them
    c. Test whether core PyBaMM does not carry optional dependencies parsed in point 2

All three of them could then be placed as regular testing methods in `test_util.py`, and they will then work on Python 3.8–3.12.

With this, we can avoid another GitHub Actions workflow or separate files that we need to keep maintaining. It might make sense to add `importlib_metadata` to our `[dev]` dependencies too, so that it will be available for both `python run-tests.py --unit` and `nox -s unit` ways of invocation for the unit tests.

Reply:
Ok now I see your point, it is a subtle difference to my view and makes perfect sense.
In 5c069cab6d7ccb299f2df1a7df8e1eb6631ce441, c1a6c891bd82737bf654025b8de4090ec9d5a3e7, c1573c8d8f9c28a0348e6d73e1a49ec7fa1ede1c I should have made all the necessary changes to eliminate additional workflows and integrate dependency testing into the actual tests (as suggested, I also added `importlib_metadata` in the `[dev]` dependencies to cope with python versions `< 3.10`). This way we should have cleaner code.